### PR TITLE
fix(classifier): Fix wonky drag when image is rotated

### LINKED ISSUE #2273: Classifier: drag-to-pan is wonky when Subject Image is rotated in Subject Viewer
## UI Bug

Package: `lib-classifier`
Affects: the Single Image Viewer (possibly Multi-Frame Viewer too), _if_ it has Rotation enabled.

On the Classifier's Single Image Viewer, users can perform a **"drag-to-pan"** interaction. (Enable Move mode on the image toolbar. Click & drag the mouse left on the subject image. The image should now move left) This ONLY works when the subject image is **not rotated.** (i.e. rotation = 0º)

- If the subject image is rotated (via the image toolbar), then the behaviour is harder to predict. e.g. dragging left might cause the image move right.
- Of note, the "press-keyboard-keys-to-pan" interaction is WORKING FINE. Pressing the Right keyboard key will always move the image to the right, no matter how it's rotated.

### Testing

- open lib-classifier with any project/workflow that uses a Single Image Viewer, e.g. https://localhost:8080/?env=production&project=1292&workflow=18372 . or I Fancy Cats works too.
- Drag ➡️  right to pan the view. Observe that the image moves to the ➡️ right, as expected.
- Rotate the image once, so it's 90º.
- Drag ➡️  right to pan the view. Observe that the image moves ⬇️ ➡️ DOWN-RIGHT.
- Rotate the image again, so it's 180º.
- Drag ➡️  right to pan the view. Observe that the image moves ⬅️ LEFT.

Additional:
- Repeat the steps above, but instead of dragging with a mouse, press the ➡️ Right Arrow key on your keyboard. Observe that the image always moves to the ➡️ right, no matter the rotation.

### Dev Notes, Short version

Analysis:

- SVGPanZoom's `onDrag(event, difference)` function (which handles "drag-to-pan", and behaves incorrectly) should learn from its sibling `onPan(dx, dy)` function (which handles "press-keyboard-keys-to-pan")
  - `onPan()` does NOT factor for the rotational transform in `<g#transform-layer>`, and works fine.
  - `onDrag()` DOES, and has bugs.
- The `onDrag()` hander is tied to the `draggable.js`'s `dragMove()` function, which relies on its sibling `getEventOffset()` function, which _converts DOM coordinates (relative to the screen) to SVG coordinates (including rotational transformations)._

Proof of concept:
- Modify `SVGPanZoom.getEventOffset()` to `return svgPoint` (i.e. without compensating for rotation in the transform later) instead of `return svgEventOffset`
- Repeat the Testing steps above. You'll notice that the "drag mouse right" action will always move the image to the right, as expected.

### Dev Notes, Long Version

_YOU SHOULDN'T BE READING THIS. This is the raw discussion I had on Slack with Jim._

I have an inkling of an idea of what's going on with the "drag-to-pan is wonky when Subject Image is rotated" issue. This issue is NOT related to the recent SVG context changes, and (if I'm correct) has existed for quuuiiite a while now without anybody noticing.

(I've time-travelled back to 28 Apr (94c750126a9cbbbe222c323987d13b2840bcecc2) and can confirm the issue exists there)

If we use Keyboard Keys (up/down/left/right) on a standard SingleImageViewer, we're triggering SVGPanZoom's onPan(dx, dy) function. No matter the rotation of the Subject Image, dx/dy will always be relative to the screen. (e.g. pressing the Right Key will result in dx=1, dy=0, whether the Subject Image is rotated or not.) Cool.

Conversely, if we use "mouse-drag-to-pan", we're triggering SVGPanZoom's onDrag(event, difference) function. Here, the value of difference.x and difference.y is VERY volatile and seems relative to the subject image('s rotation). (e.g. dragging Right will result in difference.x=+positive, difference.y = 0 only if rotation == 0. If rotation == 90º, then dragging Right will instead result in difference.x=+positive and difference.y=+positive.)

My current trace is as follows:
SingleImageViewerContainer prepares the onDrag hook -> this hook is passed to SubjectImage... -> which is a draggable.js -> draggable.js calculates the difference for the "drag-to-pan" interaction in dragMove() -> ...which uses convertEvent() -> ...which uses getEventOffset() -> which uses draggable.getEventOffset() which does use the <g#transform-layer> and factors in the rotation transform on the Subject image. And it probably shouldn't!

:exclamation: Solution: the "drag-to-move" logic should NOT need to factor for "DOM-to-SVG coordinate conversion", aka getEventOffset(), for the mouse event.

Why? Both onPan() and onDrag() perform their changes on the viewBox of top-level <svg viewbox="x0 y0 w0 h0"> , and therefore shouldn't need to factor the transorm of the one-level-down <g#transform-layer> . (Ugh, I'm still trying to digest the idea of that sentence.) But in simpler terms: 1. onPan and onDrag are performing the same op on svg.viewbox. 2. onPan works, onDrag doesn't. 3. therefore, onDrag should follow what onPan does, i.e. always be relative to the screen, so pressing the Right key or dragging to the right must result in dx/difference.x = +positive, dy/difference.y = 0

Simple test: modify draggable.getEventOffset() to return svgPoint instead of return svgEventOffset . Skipping the matrix transform step seems to do the trick!

### Status

UI bug. Good to fix, but not urgent since users haven't raised flags on the matter yet. This WILL become more relevant when we have more Transcription project that rely on users rotating images, though.

### LINKED ISSUE #6067: Dragging subject images ignores rotation transforms
## Package
_Choose from the list:_
- [ ] app-content-pages
- [ ] app-project
- [x] lib-classifier
- [ ] lib-panoptes-js
- [ ] lib-react-components
- [ ] unknown

## Describe the bug
Dragging a rotated subject image ignores the rotation transformation, leading to weird and unexpected UX eg. I drag right but the image moves up (its RHS in the rotated frame.)

Panning with the keyboard isn't affected, only dragging with a pointer (mouse, trackpad or finger.)

https://github.com/zooniverse/front-end-monorepo/assets/59547/701789b3-32e3-4bc8-ad99-9c86ecef56ae

## To Reproduce
Zoom an image, either in or out, rotate it, and drag the rotated image to move. 

## Expected behavior
Dragging should move the image in the frame of the subject viewer, not the rotated frame of the image eg. dragging right should move the rotated image to my right. 

## Device information
Tested in desktop Firefox and Chrome,  and mobile Safari on an iPhone XS. This seems to be a bug in the Zooniverse SVG code, not a browser bug. 

## Additional context
Dragging worked correctly when rotation was first introduced, for transcription projects.

### PR DESCRIPTION
**Pull request by eatyourgreens**
- Refactor `SVGPanZoom` to use a ref for `zoom`, rather than component state. Hopefully, this will lead to fewer rerenders when dragging a zoomed image, and smoother performance.
- Use a ref to store pointer coordinates in the `draggable` decorator, to avoid unnecessary rerenders.
- Move the cropped SVG viewbox into a nested SVG, for better performance when dragging a zoomed image.
- Remove the `1.5` easing factor on dragged images, so that they 'stick' to the pointer when dragged.
- Use `focus-visible` so that focus rings are only shown for keyboard interactions.
- refactor the `useKeyZoom` hook to support rotated images. 

Screen recordings from Firefox 125 running on an M3 MacBook (Sonoma 14.4.1.)

Pan and zoom with a trackpad:

https://github.com/zooniverse/front-end-monorepo/assets/59547/43393ea6-3cd2-4b13-9e19-afef14865b58

Pan and zoom with a trackpad at a very high zoom level:

https://github.com/zooniverse/front-end-monorepo/assets/59547/bfe58619-dc91-4fe3-a611-07c2fc7bbaab

Keyboard focus with Tab and shift-Tab:

https://github.com/zooniverse/front-end-monorepo/assets/59547/bc92375e-c3e1-44a3-b70f-b9291e45081a

## Package
- lib-classifier

## Linked Issue and/or Talk Post
- fixes #2273.
- fixes #6067.

## How to Review
Zoom and pan an image in the dev classifier, using both the keyboard and a trackpad/mouse to pan the zoomed image. I've been testing this with Wildwatch Burrowing Owl in Chrome, Safari and Firefox.
https://localhost:8080/?project=sandiegozooglobal/wildwatch-burrowing-owl&env=production

You can test the focus styling by tabbing through the page, to check that drawn marks have focus styling when focused ([WCAG Success Criterion 2.4.7: focus visible](https://www.w3.org/WAI/WCAG21/Understanding/focus-visible.html).) When a mark has keyboard focus, Enter will open any associated subtask popup, which should also be fully keyboard accessible. Backspace will delete the focused mark ([WCAG Success Criterion 2.1.1: Keyboard](https://www.w3.org/WAI/WCAG21/Understanding/keyboard.html).)

# Checklist
_PR Creator - Please cater the checklist to fit the review needed for your code changes._
_PR Reviewer - Use the checklist during your review. Each point should be checkmarked or discussed before PR approval._

## General
- [ ] Tests are passing locally and on Github
- [ ] Documentation is up to date and changelog has been updated if appropriate
- [ ] You can `yarn panic && yarn bootstrap` or `docker-compose up --build` and FEM works as expected
- [ ] FEM works in all major desktop browsers: Firefox, Chrome, Edge, Safari (Use Browserstack account as needed)
- [ ] FEM works in a mobile browser

## General UX
Example Staging Project: [i-fancy-cats](https://local.zooniverse.org:3000/projects/brooke/i-fancy-cats)
- [ ] All pages of a FEM project load: Home Page, Classify Page, and About Pages
- [ ] Can submit a classification
- [ ] Can sign-in and sign-out
- [ ] The component is accessible
  - Can be used with a screen reader [BBC guide to VoiceOver](https://bbc.github.io/accessibility-news-and-you/assistive-technology/testing-steps/voiceover-mac.html)
  - Can be used from the keyboard [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - It is passing accessibility checks in the storybook

## Bug Fix
- [ ] The PR creator has listed user actions to use when testing if bug is fixed
- [ ] The bug is fixed
- [ ] Unit tests are added or updated

### CODE REVIEW

**Comment by eatyourgreens** [comment]
NB. I'm testing this (and recording the screen video) on a very-much overpowered M3 MacBook Pro. It'd be useful to get tests from machines that are more in line with what typical volunteers use.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SingleImageViewer/SingleImageViewer.js [hunk 1]
```diff
@@ -66,16 +66,18 @@ function SingleImageViewer({
...
+              {enableInteractionLayer && (
+                <InteractionLayer
+                  frame={frame}
+                  height={height}
+                  scale={scale}
+                  subject={subject}
+                  width={width}
+                />
+              )}
+            </svg>
```
Comment (eatyourgreens) [inline packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SingleImageViewer/SingleImageViewer.js:80]: This nested SVG fixes #2273, #5110, and #6067.

**Comment by eatyourgreens** [comment]
Noting that the image is cropped if you try to drag it outside a rotated view box, even though there’s visibly available space in the browser viewport. I’m not sure if that would be a blocker. 

To reproduce that, rotate an image by 90° or 270°, then drag it left or right. The width of the rotated view box (the actual height of the image) is narrower than the width of the subject viewer.

**Comment by eatyourgreens** [comment]
Here's a screen recording from Chrome 124, also running on an M3 MacBook.

https://github.com/zooniverse/front-end-monorepo/assets/59547/566ab296-2be4-4409-91ce-4577f03ac1d0

File: packages/lib-classifier/src/plugins/drawingTools/models/marks/Ellipse/Ellipse.js [hunk 2]
```diff
@@ -29,8 +29,8 @@ const EllipseModel = types
 
     deleteButtonPosition(scale) {
       const theta = DELETE_BUTTON_ANGLE * (Math.PI / 180)
-      const dx = ((self.rx + BUFFER) / scale) * Math.cos(theta)
-      const dy = ((self.ry + BUFFER) / scale) * Math.sin(theta)
+      const dx = (self.rx + BUFFER) * Math.cos(theta)
+      const dy = (self.ry + BUFFER) * Math.sin(theta)
```
Comment (goplayoutside3) [inline packages/lib-classifier/src/plugins/drawingTools/models/marks/Ellipse/Ellipse.js:33]: Why is this mark touched in this PR?

**Review by goplayoutside3 — COMMENTED** [review]
While this PR does fix https://github.com/zooniverse/front-end-monorepo/issues/2273 and https://github.com/zooniverse/front-end-monorepo/issues/6067, it does not completely close https://github.com/zooniverse/front-end-monorepo/issues/5110. The image still lags behind my cursor when running this branch locally. For instance, when running i-fancy-cats, I can drag from one cat ear to the other ear as reference, yet the image does not follow at the same distance.

Just an interesting note -- the refresh rate of my display affects the janky behavior of drag. Here's a video example of me dragging a Daily Minor Planet subject on a 2021 Macbrook Pro M1 while refresh rate is 60Hz. By changing the display refresh rate to ProMotion, drag is much less janky. The same can be reproduced on production.

https://github.com/user-attachments/assets/1ac51e08-96eb-4512-aa5d-9909f3559e22

Would you like me to review this as just a fix for the wonky drag-with-rotation?

File: packages/lib-classifier/src/plugins/drawingTools/models/marks/Ellipse/Ellipse.js (hunk 2 above)
Comment (eatyourgreens) [reply packages/lib-classifier/src/plugins/drawingTools/models/marks/Ellipse/Ellipse.js:33]: There's a bug in the ellipse tool, where the delete button position changes when you zoom in and out. This fixes it.

**Comment by eatyourgreens** [comment]
>Would you like me to review this as just a fix for the wonky drag-with-rotation?

That sounds good. I see smoother dragging in Firefox on this branch, but I'm not sure about Chrome or Safari. It feels better than production, but still not great. I haven't tried experimenting with the refresh rate.

**Comment by eatyourgreens** [comment]
Here's a screen recording from Chrome on my M3 MacBook. I think the image dragging performance is still dependent on the hardware that's running the code, as this video is much smoother than yours.


https://github.com/user-attachments/assets/ff4fb899-4bec-42a3-a41b-d3a71832bf2c

**Review by goplayoutside3 — APPROVED** [review]
